api/query/search.go

```diff
@@ -85,82 +85,112 @@ func (sh structuredSearchHandler) ServeHTTP(w http.ResponseWriter, r *http.Reque
 		return
 	}
 
-	// Check if the query is a simple (empty/just True, or test name only) query
+	// Prepare logging.
+	ctx := sh.api.Context()
+	logger := shared.GetLogger(ctx)
+
 	var simpleQ TestNamePattern
+
+	r2 := r.Clone(r.Context())
+	r2url := *r.URL
+	r2.URL = &r2url
+	r2.Method = http.MethodGet
+	q := r.URL.Query()
+	q.Add("q", simpleQ.Pattern)
+	// Assemble list of run IDs for later use.
+	runIDStrs := make([]string, 0, len(rq.RunIDs))
+	for _, id := range rq.RunIDs {
+		runID := strconv.FormatInt(id, 10)
+		q.Add("run_id", runID)
+		runIDStrs = append(runIDStrs, strconv.FormatInt(id, 10))
+	}
+	runIDsStr := strings.Join(runIDStrs, ",")
+	r2.URL.RawQuery = q.Encode()
+
+	// Check if the query is a simple (empty/just True, or test name only) query
 	var isSimpleQ bool
 	{
 		if _, isTrueQ := rq.AbstractQuery.(True); isTrueQ {
 			isSimpleQ = true
 		} else if exists, isExists := rq.AbstractQuery.(AbstractExists); isExists && len(exists.Args) == 1 {
 			simpleQ, isSimpleQ = exists.Args[0].(TestNamePattern)
 		}
-		q := r.URL.Query()
 		for _, param := range []string{"interop", "subtests", "diff"} {
 			val, _ := shared.ParseBooleanParam(q, param)
 			isSimpleQ = isSimpleQ && (val == nil || !*val)
 		}
-	}
 
-	if !isSimpleQ {
-		ctx := sh.api.Context()
-		hostname := sh.api.GetServiceHostname("searchcache")
-		// TODO: This will not work when hostname is localhost (http scheme needed).
-		fwdURL, _ := url.Parse(fmt.Sprintf("https://%s/api/search/cache", hostname))
-		fwdURL.RawQuery = r.URL.RawQuery
-
-		logger := shared.GetLogger(ctx)
-		logger.Infof("Forwarding structured search request to %s: %s", hostname, string(data))
-
-		client := sh.api.GetHTTPClientWithTimeout(time.Second * 15)
-		req, err := http.NewRequest("POST", fwdURL.String(), bytes.NewBuffer(data))
-		if err != nil {
-			logger.Errorf("Failed to create request to POST %s: %v", fwdURL.String(), err)
-			http.Error(w, "Error connecting to search API cache", http.StatusInternalServerError)
-			return
+		// Check old summary files. If any can't be found,
+		// use the searchcache to aggregate the runs.
+		summaryErr := sh.validateSummaryVersions(r2.URL.Query(), logger)
```

Comment:
this becomes:

```go
// Check old summary files. If any can't be found,
// use the searchcache to aggregate the runs.
err = sh.validateSummaryVersions(r2.URL.Query(), logger)
if err != nil {
   switch err {
   case errors.Is(err, ErrBadSummaryVersion):
      logger.Warnf("%s yields unsupported summary version. %s", r2.URL.Query().Encode(), err.Error()) // Maybe this should be Debugf?
      isSimpleQ = false
   default:
      logger.Debugf("Error checking summary file names: %v", err)
      // for all other errors, do we want to do nothing to simpleQ?
    }

}

```

Reply:
Originally, the method returned false if any error occurred, so we'll set `isSimpleQ` to false if any error is present.

Also, I was technically having issues here because `errors.Is()` returns a boolean, which is not consistent with comparing against the error, so I switched this to an if-else block.

Reply:
you're right. in my example you would need to change the line `switch err` to `switch` (remove the variable to compare). Using an if-statement is completely fine too!